gridData/mrc.py

```diff
@@ -0,0 +1,136 @@
+# gridDataFormats --- python modules to read and write gridded data
+# Copyright (c) 2009-2021 Oliver Beckstein <orbeckst@gmail.com>
+# Released under the GNU Lesser General Public License, version 3 or later.
+#
+
+""":mod:`mrc` --- the mrc/CCP4 volumetric data format
+===================================================
+
+.. versionadded:: 0.7.0
+
+Reading of MRC/CCP4 volumetric files (`MRC2014 file format`_) using
+the mrcfile_ library [Burnley2017]_.
+
+This implementation replaces the :mod:`CCP4` module.
+
+.. _mrcfile: https://mrcfile.readthedocs.io/
+.. _`MRC2014 file format`: http://www.ccpem.ac.uk/mrc_format/mrc2014.php
+
+
+References
+----------
+
+.. [Burnley2017] Burnley T, Palmer C and Winn M (2017) Recent
+                 developments in the CCP-EM software suite. *Acta
+                 Cryst.* D73:469-477. doi: `10.1107/S2059798317007859`_
+
+
+.. _`10.1107/S2059798317007859`: https://doi.org/10.1107/S2059798317007859
+
+Classes
+-------
+
+"""
+from __future__ import absolute_import
+from six.moves import range
+
+import warnings
+
+import numpy as np
+import mrcfile
+
+
+class MRC(object):
+    """Represent a MRC/CCP4 file.
+
+    Load `MRC/CCP4 2014 <MRC2014 file format>`_ 3D volumetric data with
+    the mrcfile_ library.
+
+    Parameters
+    ----------
+    filename : str (optional)
+       input file (or stream), can be compressed
+
+
+    Attributes
+    ----------
+    header : numpy.recarray
+       Header data from the MRC file as a numpy record array.
+
+    array : numpy.ndarray
+       Data as a 3-dimensional array where axis 0 corresponds to X,
+       axis 1 to Y, and axis 2 to Z. This order is always enforced,
+       regardless of the order in the mrc file.
+
+    delta : numpy.ndarray
+       Diagonal matrix with the voxel size in X, Y, and Z direction
+       (taken from the :attr:`mrcfile.mrcfile.voxel_size` attribute)
+
+    origin : numpy.ndarray
+       numpy array with coordinates of the coordinate system origin
+       (taken from :attr:`header.origin`)
+
+    rank : int
+       The integer 3, denoting that only 3D maps are read.
+
+
+    Notes
+    -----
+    * Only volumetric (3D) densities are read.
+    * Only orthorhombic unitcells supported (other raise :exc:`ValueError`)
+    * Only reading is currently supported.
+
+
+    .. versionadded:: 0.7.0
+
+    """
+
+    def __init__(self, filename=None):
+        self.filename = filename
+        if filename is not None:
+            self.read(filename)
+
+    def read(self, filename):
+        """Populate the instance from the MRC/CCP4 file *filename*."""
+        if filename is not None:
+            self.filename = filename
+        with mrcfile.open(filename) as mrc:
+            if not mrc.is_volume():                           #pragma: no cover
+                raise ValueError(
+                    "MRC file {} is not a volumetric density.".format(filename))
+            self.header = h = mrc.header.copy()
```

Comment:
any difference between header and extended_header here?

Reply:
don't know — experts??

Reply:
From https://www.ccpem.ac.uk/mrc_format/mrc2014.php, 

> The extended header is now used by different software to hold various additional metadata instead, as indicated by the EXTTYP tag. 

The `EXTTYP` flag takes seven different values. You could make a copy incase the user wants it.  Note that `self.header` does not seem to be accessible from `Grid` or, at least, it is not obvious to me how to access it.

Reply:
Do you want to have access to `header` ?

Reply:
I don't have a personal-use case in mind.  But, if it isn't difficult, it could be beneficial in cases we haven't thought of yet. Not everything in the `header` ends up in a `Grid` object.

Reply:
I am adding the header as undocumented `Grid._mrc_header`. If it is used then we can think of a better way to keep format-specific data.